src/directives/offense/poisonRoom.ts

```diff
@@ -1,169 +1,188 @@
-// 		}
-// 	}
-// }
-//
-//
-//
-//
-//
+import {Colony, ColonyMemory, getAllColonies} from '../../Colony';
+import {log} from '../../console/log';
+import {ClaimingOverlord} from '../../overlords/colonization/claimer';
+import {RoomPoisonerOverlord} from '../../overlords/offense/roomPoisoner';
+import {StationaryScoutOverlord} from '../../overlords/scouting/stationary';
+import {profile} from '../../profiler/decorator';
+import {Cartographer, ROOMTYPE_CONTROLLER} from '../../utilities/Cartographer';
+import {printRoomName} from '../../utilities/utils';
+import {MY_USERNAME} from '../../~settings';
+import {DirectiveOutpostDefense} from '../defense/outpostDefense';
+import {Directive} from '../Directive';
+import {DirectiveControllerAttack} from './controllerAttack';
+
+export const RUN_TIMER = 25;
+
+/*
+* Poison sources in remote rooms by claiming controller, walling in its sources and controller then unclaiming it.
+*/
+@profile
+export class DirectivePoisonRoom extends Directive {
+	static directiveName = 'PoisonRoom';
+	static color = COLOR_RED;
+	static secondaryColor = COLOR_BROWN;
+	static requiredRCL = 4;
+	walkableSourcePosisions: RoomPosition[];
+	walkableControllerPosisions: RoomPosition[];
+	overlords: {
+		claim: ClaimingOverlord;
+		roomPoisoner: RoomPoisonerOverlord;
+	};
+	constructor(flag: Flag) {
+		super(flag, colony => colony.level >= DirectivePoisonRoom.requiredRCL);
+		// Remove if misplaced
+		if (Cartographer.roomType(this.pos.roomName) != ROOMTYPE_CONTROLLER) {
+			log.warning(`${this.print}: ${printRoomName(this.pos.roomName)} is not a controller room; ` +
+						`removing directive!`);
+			this.remove(true);
+		}
+		// failsafe - if already owned and controller level > 2, remove flag
+		if(this.room && this.room.controller && this.room.controller.my && this.room.controller.level > 2) {
+			this.remove(true);
+		}
+		// Remove if owned by other player
+		if (this.room && !!this.room.owner && this.room.owner != MY_USERNAME) {
+			log.notify(`Removing poisonRoom directive in ${this.pos.roomName}: room already owned by another player.`);
+			this.remove(true);
+		}
+		// Remove if not enough GCL
+		if(getAllColonies().length == Game.gcl.level) {
+			log.warning(`${this.print}: ${printRoomName(this.pos.roomName)} not enough GCL; ` +
+						`for poison room, removing directive!`);
+		}
+	}
+	spawnMoarOverlords() {
+		this.overlords.claim = new ClaimingOverlord(this);
+		this.overlords.roomPoisoner = new RoomPoisonerOverlord(this);
+	}
+	init() {
+		this.alert(`Poisoning Room ${this.pos.roomName}`);
+
+		if(!(this.room && this.room.controller)) {
+			return;
+		}
+
+		// suspend room immediately once claimed to avoid default room actions
+		if(this.room.controller.level == 1) {
+			this.suspendColony(this.room.name);
+		}
+
+		if (Game.time % RUN_TIMER == 0 || this.room.constructionSites.length == 0) {
+			// capture variables
+			this.walkableSourcePosisions	 = _.filter(_.flatten(_.map(this.room.sources,
```

Comment:
how about you just use 
```
const thingsToBlock = _.compact([...this.room.sources, this.room.controller, this.room.mineral]);
this.blockPositions = _(thingsToBlock).map(obj => obj.pos).unique(pos => pos.name).filter(pos => pos.isWalkable)
```
This also fixes a bug which you will encounter if you have two adjacent sources where you'll end up with duplicate neighbors.

Reply:
I need help, tried to implement this but it is giving me an error.
do not know how this chaining is working in order to fix it!

get positionsToBlock(): RoomPosition[] {
		if (!this.room) {
			return [];
		}
		const thingsToBlock = _.compact([this.room.controller, ...this.room.sources]) as RoomObject[];
		return _(thingsToBlock).map(obj => obj.pos).unique(pos => pos.name).filter(pos => pos.isWalkable);
	}

error: 
Type 'LoDashImplicitArrayWrapper<RoomPosition>' is missing the following properties from type 'RoomPosition[]': length, reduceRight, copyWithin, [Symbol.iterator], and 6 more.ts(2740)